### PR TITLE
Fix pylint parallel execution

### PR DESCRIPTION
**Pull request by koniiiik**
## Description

This patch prevents closing over the entire `ProspectorConfig` in the filter generated by `make_exclusion_filter` to avoid pickling the entire object when pylint forks subprocesses for its parallel execution.

## Related Issue

#573 

## Motivation and Context

Full context explained in https://github.com/PyCQA/prospector/issues/573#issuecomment-1380048719

This branch also includes a slight bump of the lower bound on Python in pyproject.toml. The reason is that allowing Python 3.7.1 locks pylint to versions below 2.14, and pylint 2.13 has an even more catastrophic failure mode than 2.15 – instead of crashing, it livelocks in an endless loop of spawning new subprocesses.

Lifting the Python lower bound to 3.7.2 allows poetry to upgrade pylint to 2.15, which handles the failure state in a slightly more graceful way.

## How Has This Been Tested?

Ran pytest on Python 3.10. Also makes it possible for prospector to run successfully on our code base at $DAYJOB.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] My change requires a change to the dependencies
- [x] I have updated the dependencies accordingly
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -38,7 +38,7 @@ include = [
 prospector = 'prospector.run:main'
 
 [tool.poetry.dependencies]
-python = ">=3.7,<4.0"
+python = ">=3.7.2,<4.0"
```
Comment (carlio) [inline pyproject.toml:41]: Is this change necessary for this fix? I imagine it won't affect too much but I'd like to avoid changing python comparability ranges in a bugfix release.

**Review by carlio — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (koniiiik) [reply pyproject.toml:41]: Explained in the “Motivation and Context” section of the PR description. Not necessary for the fix itself, but necessary for the tests to not hang indefinitely without the fix. I don't have a strong opinion either way, it feels a bit like a “damned if you do, damned if you don't” kind of situation.

A case could be made that anyone running Python 3.7.0 or 3.7.1 has bigger issues on their hand than losing official support for the latest prospector, and there are also high-profile projects with an official policy of only supporting the latest patch level of each of the respective supported minor branches of Python.

Ultimately, this change was really kinda forced by poetry, because it really *really* didn't want to upgrade to pylint 2.15 without it, but nothing will break in an already working state if I roll it back, so if you prefer me to do that, then can do.

**Comment by koniiiik** [comment]
> Another question, is the `two.py` test file intentionally blank? I wasn't sure.

Oh, yeah, that's rather arbitrary, I created multiple files right off the bat, because I wasn't sure if parallel execution would fully trigger without them, but I might be able to clean it up.

File: pyproject.toml (hunk 1 above)
Comment (carlio) [reply pyproject.toml:41]: Sorry I had not properly read the motivation part. I think you're right that we should update it due to the not being able to have the newer pylint versions.

Prospector really tries to be as vague as possible about version ranges to maximise compatibility with other code linting tools which might be run separately in different contexts, eg IDE plugins which run pylint or whatever. However at some point the legacy code becomes too crufty and leads to further issues. I am maybe just being too paranoid!

Especially since you're right, people stuck on python 3.7.1 then prospector shouldn't hold more up-to-date users to ransom with an older pylint.

I might suggest a `1.9` release instead of a `1.8.5` in this case, what do you think @Pierre-Sassoulas ? I'm easy either way.

**Comment by carlio** [comment]
> Oh, yeah, that's rather arbitrary, I created multiple files right off the bat, because I wasn't sure if parallel execution would fully trigger without them, but I might be able to clean it up.

It's not an issue to leave it, it just seemed an anomaly so wanted to check!

**Comment by koniiiik** [comment]
> It's not an issue to leave it, it just seemed an anomaly so wanted to check!

Turns out it's not necessary, so in the interest of avoiding any potential confusion, I cleaned it up.

**Review by Pierre-Sassoulas — APPROVED** [review]
I think upgrading to python 3.7.2 is not that big of a deal, some tool like flake8 already dropped python 3.7. And the EOL for the latest python 3.7 is very soon anyway.

**Review by carlio — APPROVED** [review]
